### PR TITLE
Get executable path

### PR DESCRIPTION
**Pull request by Ryno95**
So I got over excited rebuilding the functionality to exclude as many mallocs as possible. First time doing it this way so please send some tips wherever possible!

The two functions for use is get_executable_path() and is_executable():

get_executable_path:
- Just pushed and realised calloc needs to be changed to ft_calloc :(
- There is an error check on line 49, not sure if it's necessary?

is_executable:
- Basically a stat() call in a return, name is just nicer hehe :) 
- Can't decide if i want it t_bool or not...
- Testing for is_executable is kinda difficult as I can only run it against things that have already been used
- Making it t_bool i can test against expected output, but it will differ from system to system

### CODE REVIEW

File: src/parser/get_executable_path.c [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+char	*get_executable_path(const char *command)
+{
+	const char	*all_paths = getenv(PATH);
+	char		*executable_buffer;
+	int			single_path_len;
+	struct stat	status;
+
+	executable_buffer = calloc(CMD_BUFFER_SIZE, sizeof(char));
```
Comment (momeyer) [inline src/parser/get_executable_path.c:42]: CMD_BUFFER_SIZE should be bigger to prevent overflow, I'd assume something like 4096.
Instead of using calloc, on line 38 you can do:
`char exectable_buffer[CMD_BUFFER_SIZE];`
This way the buffer will be in the stack and the memory will be freed automatically at the end of the function scope, so we can have a buffer with a bigger size and we don't need to check if the calloc failed, otherwise, in line 43 for example if !command we still have to free the buffer, same before returning on line 56 before returning, with the array we don't need to worry about it. I know you return the buffer, we can call` ft_strdup(buffer)` in the return statement on line 53, so it will malloc just the memory that is necessary, and return NULL in case of failure.
I think we have to "cleanup" buffer every time we start the while loop, right? So maybe something like this:
`ft_bzero(&executable_buffer[0], CMD_BUFFER_SIZE);`
this syntax `&executable_buffer[0]` can be used in this case instead of casting to `char *`.
for the other functions you can still have the argument as `char *` and then you pass the buffer the same way as in bzero.

File: src/parser/get_executable_path.c [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+	if (!executable_buffer || !command || !all_paths)
+		return (NULL);
+	while (*all_paths)
+	{
+		single_path_len = get_sub_path_len(all_paths);
+		get_single_path(all_paths, executable_buffer, single_path_len);
+		if (single_path_len != (int)ft_strlen(executable_buffer))
```
Comment (momeyer) [inline src/parser/get_executable_path.c:49]: Don't think this check is necessary

File: src/parser/get_executable_path.c [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+{
+	int	i;
+
+	i = 0;
+	while (path && path[i] != '\0' && path[i] != COLON)
+		i++;
+	return (i);
+}
+
+static void	get_single_path(const char *all_paths, char *buffer, int size)
```
Comment (momeyer) [inline src/parser/get_executable_path.c:25]: I think the name could be something like `copy_executable_path_to buffer`

File: src/parser/get_executable_path.c (hunk 1 above)
Comment (Ryno95) [reply src/parser/get_executable_path.c:42]: Believe it or not but but thats how i had it first hehe
I couldn't figure out a way to return the path_string, very nice trick with the str_dup!

File: src/parser/get_executable_path.c (hunk 2 above)
Comment (Ryno95) [reply src/parser/get_executable_path.c:49]: Cool, bye Felicia! :)

File: src/parser/get_executable_path.c (hunk 3 above)
Comment (Ryno95) [reply src/parser/get_executable_path.c:25]: mmmm don't fully agree with this as the path is not executable, that is what we are checking after we've gotten the single_path, I was struggling to give this a proper name, maybe `copy_single_path_to_buffer` ?

**Comment by Ryno95** [comment]
Implemented suggested changes, buffer_array instead of malloced buffer and returning srtdup

**Review by momeyer — APPROVED** [review]
Looks good!
